### PR TITLE
Start powering the spreadsheet from organizer columns

### PR DESCRIPTION
**Pull request by bhollis**
This is maybe overkill, but it's something I've wanted to do for a long time. With this change, the inventory spreadsheets are entirely driven by the column definitions from the Organizer. This wasn't quite as smooth as I had imagined, because I wanted to preserve the output of the spreadsheets as they were, which required a bit more custom stuff than I had wanted. For example, every column needs a custom non-localized name, and the order of columns needs to be handled in a separate list. However, my goal was to make it so you can't accidentally add a column to Organizer without it appearing in Spreadsheets and vice versa, and that goal was achieved.

Some actual changes:
1. The CSVs are sorted by item name now, instead of whatever order from inventory.
2. Ghosts have extra columns - year/season and energy.
3. Added a "Foundry" column to the Organizer.
4. Removed "Guard Efficiency" stat which was deprecated.
5. The output file names are slightly different (destiny-weapon.csv vs destinyWeapons.csv)
6. A few other minor things I'll remember as I re-review the code.

### CODE REVIEW

File: src/app/organizer/table-types.ts [hunk 1]
```diff
@@ -52,6 +54,27 @@ export interface ColumnDefinition<V extends Value = Value> {
    * but sometimes a custom stat should be limited to only displaying for a certain class
    */
   limitToClass?: DestinyClass;
+
+  /**
+   * A name for this column when it is output as CSV. This will reuse the value
+   * function as-is. We could reuse the header, but that's localized, while
+   * historically our CSV column names haven't been. I tried to combine this and
+   * `csvVal` but ran into type covariance issues.
```
Comment (bhollis) [inline src/app/organizer/table-types.ts:62]: Basically, if you do:
```ts
  csv: string | ((value: V, item: DimItem, spreadsheetContext: SpreadsheetContext) => CSVColumn | CSVColumn[]);
```

Then TypeScript will complain that, for example, `ColumnDefinition<string>` can't assign a `((value: string, item: DimItem, spreadsheetContext: SpreadsheetContext) => CSVColumn | CSVColumn[])` to `((value: Value, item: DimItem, spreadsheetContext: SpreadsheetContext) => CSVColumn | CSVColumn[])`

File: src/app/organizer/Columns.tsx [hunk 2]
```diff
@@ -267,35 +299,52 @@ export function getColumns(
...
+        csv: 'Hash',
+        value: (i) => i.hash,
+      }),
+    isSpreadsheet &&
+      c({
+        id: 'id',
+        header: 'Id',
+        csv: 'Id',
+        value: (i) => `"${i.id}"`,
+      }),
```
Comment (bhollis) [inline src/app/organizer/Columns.tsx:343]: I had to add a few spreadsheet-only columns.

File: src/app/organizer/Columns.tsx [hunk 3]
```diff
@@ -588,9 +706,11 @@ export function getColumns(
         header: t('Organizer.Columns.Level'),
         value: (item) => item.craftedInfo?.level,
         defaultSort: SortDirection.DESC,
+        csvVal: (value) => ['Crafted Level', value ?? 0],
```
Comment (bhollis) [inline src/app/organizer/Columns.tsx:709]: In general I want columns to return `undefined` for their value if the column doesn't apply. But then annoyingly the CSV was inconsistent about whether a column was undefined or zeroed.

File: src/app/organizer/Columns.tsx [hunk 4]
```diff
@@ -960,3 +1097,55 @@ function getIntrinsicSockets(item: DimItem) {
     ? [intrinsicSocket, ...extraIntrinsicSockets]
     : extraIntrinsicSockets;
 }
+
+/**
+ * This builds stat infos for all the stats that are relevant to a particular category of items.
+ * It will return the same result for the same category, since all items in a category share stats.
+ */
+export function buildStatInfo(items: DimItem[]): {
```
Comment (bhollis) [inline src/app/organizer/Columns.tsx:1105]: Moved from `ItemTable.tsx`

File: src/app/organizer/table-types.ts (hunk 1 above)
Comment (robojumper) [reply src/app/organizer/table-types.ts:62]: You can use the good old bivariance hack for this if you really want to combine them:
```
type CSVOutput<V> = {
  bivarianceHack(
    value: V,
    item: DimItem,
    spreadsheetContext: SpreadsheetContext,
  ): CSVColumn | CSVColumn[];
}['bivarianceHack'];
```
then `csv?: string | CSVOutput<V>;` works

Comment (bhollis) [reply src/app/organizer/table-types.ts:62]: Aha! I knew there was a hack like this, but couldn't remember it.

File: src/app/organizer/table-types.ts [hunk 5]
```diff
@@ -52,6 +54,33 @@ export interface ColumnDefinition<V extends Value = Value> {
...
```
Comment (robojumper) [inline src/app/organizer/table-types.ts:62]:
```suggestion
   * historically our CSV column names haven't been.
```

**Pull request merged**